enhancements/olm/max-openshift-versions-for-operators.md

```diff
@@ -0,0 +1,284 @@
+---
+title: Allow OLM Managed Operators to Specify a Max OpenShift Version
```

Comment:
Hi @awgreene, 

I know that this EP is merged already. However, it only came to my attention now.
I raised some concerns/questions over it in the channel which might be valid we address here. 

1. We have the OCP label `LABEL com.redhat.openshift.versions: v4.5-v4.7` that can be used in the image index already. What happens if a user adds the `label com.redhat.openshift.versions` in the image and the `maxOCPversion` annotation in the CSV? Also, as described in the https://github.com/openshift/enhancements/pull/592/files#r570641713 how will we ensure that the value in the annotation is compatible with minKubeversion?

- Should not CVP verify if it?
- Should not this annotation and its compatibility with `minKubeversion` be validated via the operator-framework/api and then, in this way, it is checked via `operator-sdk bundle validate` and etc? 

2. Could also we add the annotation for maxKubeVersion in order to make it compatible? Also, would not make sense deprecated the attr minKubeVersion and add it via annotation as well? And then, we have a minOCPVersion too?  

3. Would be possible to make this annotation be generic and valid for any vendor instead of being specific for openshift? 

c/c @kevinrizza @dmesser

Reply:
Hello @camilamacedo86 - thanks for the review!

> We have the OCP label LABEL com.redhat.openshift.versions: v4.5-v4.7 that can be used in the image index already. What happens if a user adds the label com.redhat.openshift.versions in the image and the maxOCPversion annotation in the CSV? 

Can you elaborate on this label and its impact? I believe that @gallettilance and I [discussed](https://github.com/openshift/enhancements/pull/592#discussion_r560518168) this earlier and I had some concerns with overloading the existing field.

> Also, as described in the https://github.com/openshift/enhancements/pull/592/files#r570641713 how will we ensure that the value in the annotation is compatible with minKubeversion? Should not CVP verify if it?

There is no guarantee that these annotation will not conflict. If either of these constraints are not met the operator will not be installed.

> Should not this annotation and its compatibility with minKubeversion be validated via the operator-framework/api and then, in this way, it is checked via operator-sdk bundle validate and etc?

Again, these are independent constraints. Also, if this approach was pursued, are you proposing that the validation library hardcodes the relationship between OCP and K8s version?

> Could also we add the annotation for maxKubeVersion in order to make it compatible?

This could certainly be added.

> Also, would not make sense deprecated the attr minKubeVersion and add it via annotation as well? And then, we have a minOCPVersion too?

We could remove this field from existing CRDs but I do not believe that it is worth doing so until we release a new version of the CRD as existing CRs would need to be migrated or OLM would need to ship a conversion webhook.

> Would be possible to make this annotation be generic and valid for any vendor instead of being specific for openshift?

While I agree with this in principle, I am not sure how users could provide specific versions for different Kubernetes variations without unique annotations. For example, how could an Operator Author generically specify that their operator's `maxVersion` is `4.7` for openshift but `1.21` for vanilla k8s?

Reply:
Hi @awgreene, 

Sorry, if I could not clarifies. My mainly questions/concerns are about:

**Is it aligned with what exists already or are we planning to change what exists to be aligned with?** 

Otherwise, how it works and how it should be used can end up bring confusion and not to be intuitive to the end-users and us indeed. (e.g skipRange which does not work exactly as skips )  

**How users will know that they did not make something wrong? How we will know what should be the expected behaviour? How can we prevent an operator be published with an OCP label and max version or minKubeversioon that is not compatible and will never be installed?**

Example A:
- LABEL com.redhat.openshift.versions: v4.5-v4.7
- maxOCPversion=4.6 

Will the operator be installed or not in 4.7? 

Example B:
- LABEL com.redhat.openshift.versions: v4.5-v4.7
- maxOCPversion=4.8

Will the operator be installed in 4.8 or not? 

- And then, will the `maxOCPversion` prevent the operator be in the index catalogue as it is done for skips or will it works as skipRange? 
- How the users will know that thy added an invalid value for this field before publish and check that the operator is not installed? How can we prevent this scenario? Would be possible to add a check to the bundle validate for that? 

**Could we provide a solution that is valid for upstream and downstream?**

Do we really need to create a specific annotation for OCP or could we have for example maxVendorVersion and compare that version with the vendor version one? I am not sure if it would be possible, however, did we check if we could use k8s api to get the vendor version and do the check no matter the vendor kind or this information cannot be obtained by k8s api?

Reply:
@camilamacedo86 the label decides which catalogs the bundle will be in.  the min/maxVersion field decides which clusters it can be installed on.

They are related, but independent.  There is nothing invalid about saying "put this bundle in the v4.8 catalog" and then saying "this bundle can't be installed on v4.8".  It's a poor choice by the operator team, but there's no need to explicitly forbid it. (Among other things, a cluster admin could still reference the v4.8 catalog from another cluster version which might be able to install the operator.  Also someday we're going to have the ability for admins to force the installation despite requirements not being met).


> Do we really need to create a specific annotation for OCP or could we have for example maxVendorVersion and compare that version with the vendor version one? I am not sure if it would be possible, however, did we check if we could use k8s api to get the vendor version and do the check no matter the vendor kind or this information cannot be obtained by k8s api?

what would operator teams set "maxVendorVersion" to?  Openshift has both an OCP version and a kube version.  Both versions can be retrieved and compared.

Reply:
Thank you for the clarifications.